vllm/model_executor/layers/linear.py

```diff
@@ -538,6 +635,109 @@ def __init__(self,
                          params_dtype=params_dtype,
                          quant_config=quant_config)
 
+    def _get_shard_offset_mapping(self, loaded_shard_id: str):
+        shard_offset_mapping = {
+            "q": 0,
+            "k": self.num_heads * self.head_size,
+            "v": (self.num_heads + self.num_kv_heads) * self.head_size,
+            "total": (self.num_heads + 2 * self.num_kv_heads) * self.head_size
+        }
+        return shard_offset_mapping.get(loaded_shard_id)
+
+    def _get_shard_size_mapping(self, loaded_shard_id: str):
+        shard_size_mapping = {
+            "q": self.num_heads * self.head_size,
+            "k": self.num_kv_heads * self.head_size,
+            "v": self.num_kv_heads * self.head_size,
+        }
+        return shard_size_mapping.get(loaded_shard_id)
+
+    def _default_loading(self, param: vLLMParameter, param_data: torch.Tensor,
```

Comment:
I think we should move `_default_loading` into `vLLMParameter`

Then, the function signature would look like:

```python
class vLLMParameter:
    def load_merged_column_parallel_linear(layer: MergedColumnParallelLinear, loaded_weight: torch.tensors):
          pass
    def load_row_parallel_linear(layer: RowParallelLinear, loaded_weight: torch.tensor):
         pass
    ....
```

Reply:
Then, with this update, in weight_loaderv2 instead of:


```python
if param.use_column_loading:
            param_data, loaded_weight = self._default_loading(
                param=param,
                param_data=param_data,
                loaded_weight=loaded_weight,
                loaded_shard_id=loaded_shard_id)
elif param.use_metadata_loading:  # What case is this?
            shard_size = loaded_weight.shape[0]
            shard_index = ["q", "k", "v"].index(loaded_shard_id)
            param_data = param_data.narrow(0, shard_index * shard_size,
                                           shard_size)
elif param.use_col_shard_split:
      param_data, loaded_weight = param.col_shard_splitter(
            param_data=param_data,
            loaded_weight=loaded_weight,
            shard_id=loaded_shard_id)
```

It would just be:

```python
param_data, loaded_weight = param.loaded_merged_column_parallel_linear(
       layer=self,
       loaded_weight=loaded_weight)
```

This would better encapsulate the logic and simplify the `linear.py` file

Reply:
I dont agree with moving this logic to the parameter.

The loading done is heavily dependent on the layer itself. At that point, we'll end up overloading the parameter with a lot of layer-specific logic.

Reply:
Yeah, I think the parameters should have:
- `load_merged_column_parallel_linear()`
- `load_qkv_column_parallel_linear()`
- `load_row_parallel_linear()`
- `load_column_parallel_linear()`